dcpy/data/compare.py

```diff
@@ -0,0 +1,320 @@
+        if len(comp_df) > 0:
+            comps[column] = comp_df.copy()
+
+    return KeyedTableComparison(
+        key_columns=key_columns,
+        left_only=left_only,
+        right_only=right_only,
+        columns_with_diffs=set(comps.keys()),
+        differences_by_column=comps,
+    )
+
+
+def _df_to_set_of_lists(df: pd.DataFrame) -> set[list]:
+    return set(list(df.itertuples(index=False, name=None)))  # type: ignore
+
+
+def compare_sql_columns(left: str, right: str, client: postgres.PostgresClient):
+    left_columns = set(client.get_table_columns(left))
+    right_columns = set(client.get_table_columns(right))
+
+    type_differences = {}
+
+    left_types = client.get_column_types(left)
+    right_types = client.get_column_types(right)
+
+    for column in left_columns & right_columns:
+        left_dtype = left_types.get(column, "None")
+        right_dtype = right_types.get(column, "None")
+
+        if left_dtype != right_dtype:
+            type_differences[column] = SimpleComparison[str](
+                left=left_dtype, right=right_dtype
+            )
+
+    return ColumnsComparison(
+        both=left_columns & right_columns,
+        left_only=left_columns - right_columns,
+        right_only=right_columns - left_columns,
+        type_differences=type_differences,
+    )
+
+
+def compare_sql_keyed_rows(
+    left: str,
+    right: str,
+    key_columns: list[str],
+    client: postgres.PostgresClient,
+    *,
+    ignore_columns: list[str] | None = None,
+):
+    left_columns = client.get_table_columns(left)
+    right_columns = client.get_table_columns(right)
+    assert set(key_columns).issubset(set(left_columns))
+    assert set(key_columns).issubset(set(right_columns))
+
+    columns = set(left_columns) & set(right_columns) - set(ignore_columns or [])
+    non_key_columns = columns - set(key_columns)
+    left_keys = ", ".join([f'"left"."{c}"' for c in key_columns])
+    left_keys_alias = ", ".join([f'"left"."{c}" AS "{c}_left"' for c in key_columns])
+    right_keys_alias = ", ".join([f'"right"."{c}"AS "{c}_right"' for c in key_columns])
+
+    on = " AND ".join([f'"left"."{c}" = "right"."{c}"' for c in key_columns])
+    from_clause = f'FROM {left} AS "left" FULL OUTER JOIN {right} AS "right" ON {on}'
+
+    left_only = client.execute_select_query(
+        f'SELECT {left_keys_alias} {from_clause} WHERE "right"."{key_columns[0]}" IS NULL'
+    )
+    right_only = client.execute_select_query(
+        f'SELECT {right_keys_alias} {from_clause} WHERE "left"."{key_columns[0]}" IS NULL'
+    )
+
+    comps: dict[str, pd.DataFrame] = {}
+
+    def query(column):
+        lc = f'"left"."{column}"'
+        rc = f'"right"."{column}"'
+        return f"""
+            SELECT 
+                {left_keys}, {lc} AS "left", {rc} AS "right" 
+            FROM {left} AS "left" 
+                INNER JOIN {right} AS "right"
+                ON {on}
+            WHERE NOT ({lc} = {rc})
```

Comment:
FYI, I think `IS DISTINCT FROM` gives the same results

Reply:
I always forget about that